### PR TITLE
Challenges review

### PR DESCRIPTION
**Pull request by vinisalazar**
Hi,

I'm reviewing the challenges for this lesson and making some minor improvements. Please let me know if I should create a separate issue to discuss this.

Episodes reviewed so far:

- [x] 01-run-quit.md
- [x] 02-variables.md
- [x] 03-types-conversion.md
- [x] 04-built-in.md
- [x] 05-coffee.md
- [x] 06-libraries.md
- [x] 07-reading-tabular.md
- [x] 08-data-frames.md
- [x] 09-plotting.md
- [x] 10-lunch.md
- [x] 11-lists.md
- [x] 12-for-loops.md
- [x] 13-conditionals.md
- [x] 14-looping-data-sets.md
- [x] 15-coffee.md
- [x] 16-writing-functions.md
- [x] 17-scope.md
- [x] 18-style.md
- [x] 19-wrap.md
- [x] 20-feedback.md

I hope these changes are helpful for the lesson.

Best wishes,
V

### CODE REVIEW

**Comment by vinisalazar** [comment]
I'd greatly appreciate a careful review of d52dcd5.

**Comment by vinisalazar** [comment]
ping @alee @ntmoore @souravsingh @vahtras

I'm aware this is a fairly large PR, so please know that I'm available for discussing, editing, or justifying any of the changes made.

Thank you for any assistance you can provide.

Best,
Vini

**Review by alee — COMMENTED** [review]
Thanks for the PR and all the hard work that went into reviewing all the challenges! 

As a minor aside in the future it's generally easier to review and merge smaller PRs (e.g., one per episode) vs one giant monolithic PR :sweat_smile: I understand why you kept them combined since it was part of a larger effort but the omnibus checklist to review all of these challenges can also be kept in a [linked issue](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)

**Comment by vinisalazar** [comment]
Hello @alee, thank you for the reply.

I understand. I will keep that in mind for future contributions. Would you like me to break this up into separate PRs, one for each episode? Or may I leave it as is?

Best wishes,
Vini

**Comment by alee** [comment]
Let's leave it as is as there is a considerable amount of work in here already, just wanted to point that out for future reference :+1: 

I'll do my best to review this and get it merged in the next few days - thanks!

**Comment by vinisalazar** [comment]
Thank you @alee !

File: _episodes/13-conditionals.md [hunk 1]
```diff
@@ -372,6 +372,18 @@ final velocity: 30.0
 > > print(smallest, largest)
 > > ~~~
 > > {: .language-python}
+> > It can be argued that an advantage of using this method would be explicitly checking for each value in the list.
```
Comment (alee) [inline _episodes/13-conditionals.md:375]: On further scrutiny I don't find this explanation very satisfying. The efficiency gains are likely minimal, I think the real gain is in clarity and we should emphasize that instead.

Efficiency wise, the second solution is running two passes over the list of `values`, as  `min(values)` and `max(values)` are both iterate over the entire `values` list. 

The most "efficient" method would be to iterate over `values` once, and replace the `min` and `max` calls with simpler conditional checks e.g.,

```
smallest = v if v < smallest
largest = v if v > largest
```

anyways I think we should stick with the clarity argument. Thoughts?

File: _episodes/16-writing-functions.md [hunk 2]
```diff
@@ -512,21 +519,21 @@ result of call is: None
 > density. In the model, time takes discrete values 0, 1, 2, ...
 >
 > 1. Define a function called `logistic_map` that takes two inputs: `x`, representing the current
->    population (at time _t_), and a parameter `r=1`. This function should return a value 
->    representing the state of the system (population) at time _t_+1, using the mapping function 
+>    population (at time `t`), and a parameter `r = 1`. This function should return a value 
+>    representing the state of the system (population) at time `t + 1`, using the mapping function:
 >
->    x(t+1) = r * x(t) * [1 - x(t)]
+>    `x[t + 1] = r * x[t] * [1 - x[t]]`
```
Comment (alee) [inline _episodes/16-writing-functions.md:525]: Hmm, I'm not sold on replacing the round parentheses with square brackets. The round parentheses makes it seem like we are invoking a method, but the square brackets makes it look like we are indexing a dictionary or list, and x isn't either, it's a mathematical function. Perhaps once we have access to [katex](https://github.com/carpentries/styles/pull/573/files) we could render this as a mathematical function instead...

In the meantime, how about replacing `x` here with the generic function `f`?

`f(t+1) = r * f(t) * (1 - f(t))`

File: _episodes/17-scope.md [hunk 3]
```diff
@@ -106,4 +106,12 @@ NameError: name 'temperature' is not defined
 > KeyError: 'Friday'
 > ~~~
 > {: .error}
+> > ## Solution
+> > 1. Three levels.
+> > 2. `errors_02.py`
+> > 3. `print_message`
+> > 4. Line 11
+> > 5. `KeyError`
```
Comment (alee) [inline _episodes/17-scope.md:114]: Perhaps we can also explain what a KeyError is or link to https://docs.python.org/3/library/exceptions.html#KeyError

**Review by alee — CHANGES_REQUESTED** [review]
Thanks for the comprehensive review of all of these challenges!

I've made a few notes where changes could be made, let me know what you think.

File: _episodes/13-conditionals.md (hunk 1 above)
Comment (vinisalazar) [reply _episodes/13-conditionals.md:375]: Yeah, I agree. I will update that.

File: _episodes/16-writing-functions.md (hunk 2 above)
Comment (vinisalazar) [reply _episodes/16-writing-functions.md:525]: Sure. Rendering that with Tex would be optimal, but I agree that `f(t)` looks better than `x[t]`.

**Comment by vinisalazar** [comment]
@alee thank you for the thorough review. Hopefully I've addressed all requested changes.

Best,
V

**Review by alee — APPROVED** [review]
LGTM, thanks @vinisalazar !

**Pull request merged**